pkg/extract/oracle/oracle.go

```diff
@@ -41,24 +72,482 @@ func Extract(args string, opts ...Option) error {
+					Operator: criteriaTypes.CriteriaOperatorTypeOR,
+					Criterions: func() []criterionTypes.Criterion {
+						criterions := make([]criterionTypes.Criterion, 0, len(pkgs))
+						for pkg, archs := range pkgs {
+							criterions = append(criterions, criterionTypes.Criterion{
+								Vulnerable: true,
+								Package: packageTypes.Package{
+									Name:          pkg.name,
+									Architectures: archs,
+								},
+								Affected: &affectedTypes.Affected{
+									Type:  rangeTypes.RangeTypeRPM,
+									Range: []rangeTypes.Range{{LessThan: pkg.fixedVersion}},
+									Fixed: []string{pkg.fixedVersion},
+								},
+							})
+						}
+						return criterions
+					}(),
+				},
+			})
+		}
+		return ds
+	}()
+
+	return data, nil
+}
+
+type ovalPackage struct {
+	major           string
+	ovalName        string
+	fixedVersion    string
+	modularityLabel string
+	arch            string
+}
+
+func collectPackages(criteria oracle.Criteria, tos tos) (packages, error) {
+	pkgs, err := evalCriteria(criteria, tos)
+	if err != nil {
+		return packages{}, errors.Wrapf(err, "applyCriteria")
+	}
+
+	allPkgs := packages{}
+	for _, p := range pkgs {
+		pkgs, ok := allPkgs[majorType(p.major)]
+		if !ok {
+			pkgs = perMajorPackages{}
+			allPkgs[majorType(p.major)] = pkgs
+		}
+		key := packageKey{
+			name: func() string {
+				switch p.modularityLabel {
+				case "":
+					return p.ovalName
+				default:
+					return fmt.Sprintf("%s::%s", p.modularityLabel, p.ovalName)
+				}
+			}(),
+			fixedVersion: p.fixedVersion,
+		}
+		pkgs[key] = append(pkgs[key], p.arch)
+	}
+	return allPkgs, nil
+}
+
+func evalCriteria(criteria oracle.Criteria, tos tos) ([]ovalPackage, error) {
+	// Exclude patterns that do not exists in oracle oval data and don't implement them, YAGNI.
+	// With these constraints, we can ignore Criteria.Operator (AND or OR) to extract package information.
+	switch {
+	case criteria.Operator == "OR" && len(criteria.Criterions) > 0:
+		return []ovalPackage{}, errors.Errorf("criteriaons under OR criteria MUST not exist")
+	case criteria.Operator == "OR" && len(criteria.Criterias) == 0:
+		return []ovalPackage{}, errors.Errorf("criterias under OR criteria MUST exist")
+	case criteria.Operator == "AND" && len(criteria.Criterias) > 1:
+		return []ovalPackage{}, errors.Errorf("criterias under AND-criteria MUST be < 2")
+	}
+
+	var pkgs []ovalPackage
+
+	for _, ca := range criteria.Criterias {
+		ps, err := evalCriteria(ca, tos)
+		if err != nil {
+			return []ovalPackage{}, err
+		}
+		pkgs = append(pkgs, ps...)
+	}
+
+	if err := evalCriterions(&pkgs, tos, criteria.Criterions); err != nil {
+		return []ovalPackage{}, err
+	}
+	return pkgs, nil
+}
+
+func evalCriterions(pkgs *[]ovalPackage, tos tos, criterions []oracle.Criterion) error {
+	if len(criterions) == 0 {
+		return nil
+	}
+
+	// The case for parent criterion is AND and without child criterions case. Add base (ANY) one
+	if len(*pkgs) == 0 {
+		*pkgs = append(*pkgs, ovalPackage{})
+	}
```

Comment:
The main role of `evalCriterions` seems to be to provide criterion content for each element of pkgs, but I feel that the above part contradicts this.

Reply:
Ahhh,,, difficult thing :)

Better or worth, I adopted your suggestion above to pull up "append" to evalCriteria(), this code has gone from evalCriterions.